tests/cases/misc.rs

```diff
@@ -51,3 +55,106 @@ fn test_peer() {
 
     assert!(resp.get_message().contains("127.0.0.1"), "{:?}", resp);
 }
+
+struct Counter {
+    global_counter: Arc<AtomicUsize>,
+    local_counter: UnsafeCell<usize>,
+}
+
+impl Counter {
+    fn incr(&self) {
+        unsafe {
+            let counter = self.local_counter.get();
+            let c = &mut *counter;
+            *c += 1;
+        }
+    }
+
+    fn flush(&self) {
+        unsafe {
+            let counter = self.local_counter.get();
+            let c = &mut *counter;
+            self.global_counter.fetch_add(*c, Ordering::SeqCst);
+            *c = 0;
+        }
+    }
+}
+
+impl Drop for Counter {
+    fn drop(&mut self) {
+        self.flush();
+    }
+}
+
+impl Clone for Counter {
+    fn clone(&self) -> Counter {
+        Counter {
+            global_counter: self.global_counter.clone(),
+            local_counter: UnsafeCell::new(0),
+        }
+    }
+}
+
+#[test]
+fn test_soundness() {
+    #[derive(Clone)]
+    struct CounterService {
+        c: Counter,
+    }
+
+    impl Greeter for CounterService {
+        fn say_hello(&self, ctx: RpcContext, _: HelloRequest, sink: UnarySink<HelloReply>) {
+            self.c.incr();
+            let resp = HelloReply::new();
+            ctx.spawn(
+                sink.success(resp)
+                    .map_err(|e| panic!("failed to reply {:?}", e)),
+            );
+        }
+    }
+
+    let env = Arc::new(EnvBuilder::new().cq_count(4).build());
+    let counter = Arc::new(AtomicUsize::new(0));
+    let service = CounterService {
+        c: Counter {
+            global_counter: counter.clone(),
+            local_counter: UnsafeCell::new(0),
+        },
+    };
+    let mut server = ServerBuilder::new(env.clone())
+        .register_service(create_greeter(service))
+        .bind("127.0.0.1", 0)
+        .build()
+        .unwrap();
+    server.start();
+    let port = server.bind_addrs()[0].1;
+
+    let spawn_reqs = |env| -> JoinHandle<()> {
+        let ch = ChannelBuilder::new(env).connect(&format!("127.0.0.1:{}", port));
+        let client = GreeterClient::new(ch);
+        let mut resps = Vec::with_capacity(3000);
+        thread::spawn(move || {
+            for _ in 0..3000 {
+                resps.push(client.say_hello_async(&HelloRequest::new()).unwrap());
+            }
+            future::join_all(resps).wait().unwrap();
```

Comment:
better to check each local counter is 3000.

Reply:
Why? I don't think it is related to the soundness.

Reply:
As I guess, you want to test different threads won't operate the same RC counter, right? If yes, how can you test it?

Reply:
Because `Counter::local_counter` is not `Sync` so if it's used in different thread concurrently there will be data race. Hence the sum of all `local_counter` should be less than 9000.